ngraph/core/src/dimension.cpp

```diff
@@ -93,20 +105,52 @@ bool Dimension::merge(Dimension& dst, const Dimension d1, const Dimension d2)
     {
         return false;
     }
-    dst = result;
+    std::string name;
+    if (d1 == d2 && d1.get_name() == d2.get_name())
+        name = d1.get_name();
+    dst = {result, name};
     return true;
 }
 
+std::string broadcast_dimensions_name(const Dimension& d1, const Dimension& d2)
+{
+    std::string name;
+    if (d1 == d2)
+    {
+        const auto& name_1 = d1.get_name();
+        const auto& name_2 = d2.get_name();
+        if (name_1 == name_2 || (name_2.empty() && !d1.has_constant_name()))
+            name = name_1;
+        else if (name_1.empty() && !d2.has_constant_name())
+            name = name_2;
+        else if (d1.has_constant_name() && d2.has_constant_name())
+            name = name_1; // + '&' + name_2;
+        else if (d1.has_constant_name() && !d2.has_constant_name())
+            name = name_2;
+        else if (!d1.has_constant_name() && d2.has_constant_name())
+            name = name_1;
+        return name;
+    }
+
+    const auto& one_dim = d1 == 1 ? d1 : (d2 == 1 ? d2 : -1);
+    const auto& other_dim = d1 == 1 ? d2 : (d2 == 1 ? d1 : -1); // it is not equal to 1
+    if (one_dim.is_dynamic())
+        return "";
+    return other_dim.get_name();
+}
+
 bool Dimension::broadcast_merge(Dimension& dst, const Dimension d1, const Dimension d2)
```

Comment:
overall the method is questionable -- it doesn't fully do what it states. but the logic for naming is strict.